Source/BoundaryConditions/PML_RZ.cpp

```diff
@@ -0,0 +1,252 @@
+#include <AMReX_GpuControl.H>
+#include <AMReX_GpuDevice.H>
+#include <AMReX_GpuLaunch.H>
+#include <AMReX_GpuQualifiers.H>
+#include <AMReX_IndexType.H>
+#include <AMReX_MFIter.H>
+#include <AMReX_ParmParse.H>
+#include <AMReX_RealVect.H>
+#include <AMReX_SPACE.H>
+#include <AMReX_VisMF.H>
+
+#include <algorithm>
+#include <cmath>
+#include <memory>
+#include <utility>
+
+using namespace amrex;
+
+PML_RZ::PML_RZ (const int lev, const BoxArray& grid_ba, const DistributionMapping& grid_dm,
+                const Geometry* geom, const int ncell, const int do_pml_in_domain)
+    : m_ncell(ncell),
+      m_do_pml_in_domain(do_pml_in_domain),
+      m_geom(geom)
+{
+
+    m_ok = true;
+
+    auto & Er_fp = WarpX::GetInstance().getEfield_fp(lev,0);
+    auto & Et_fp = WarpX::GetInstance().getEfield_fp(lev,1);
+    pml_E_fp[0] = std::make_unique<MultiFab>( amrex::convert( grid_ba, Er_fp.ixType().toIntVect() ), grid_dm,
+                                              Er_fp.nComp(), Er_fp.nGrow() );
+    pml_E_fp[1] = std::make_unique<MultiFab>( amrex::convert( grid_ba, Et_fp.ixType().toIntVect() ), grid_dm,
+                                              Et_fp.nComp(), Et_fp.nGrow() );
+
+    auto & Br_fp = WarpX::GetInstance().getBfield_fp(lev,0);
+    auto & Bt_fp = WarpX::GetInstance().getBfield_fp(lev,1);
+    pml_B_fp[0] = std::make_unique<MultiFab>( amrex::convert( grid_ba, Br_fp.ixType().toIntVect() ), grid_dm,
+                                              Br_fp.nComp(), Br_fp.nGrow() );
+    pml_B_fp[1] = std::make_unique<MultiFab>( amrex::convert( grid_ba, Bt_fp.ixType().toIntVect() ), grid_dm,
+                                              Bt_fp.nComp(), Bt_fp.nGrow() );
+
+    pml_E_fp[0]->setVal(0.0);
+    pml_E_fp[1]->setVal(0.0);
+    pml_B_fp[0]->setVal(0.0);
+    pml_B_fp[1]->setVal(0.0);
+
+}
+
+void
+PML_RZ::ApplyDamping(const std::array<amrex::MultiFab*,3>& E_fp,
+                     const std::array<amrex::MultiFab*,3>& B_fp,
+                     amrex::Real dt)
+{
+
+    const amrex::Real dr = m_geom->CellSize(0);
+    const amrex::Real cdt_over_dr = PhysConst::c*dt/dr;
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (Gpu::notInLaunchRegion())
+#endif
+    for ( amrex::MFIter mfi(*E_fp[0], amrex::TilingIfNotGPU()); mfi.isValid(); ++mfi )
+    {
+        amrex::Array4<amrex::Real> const& Et_arr = E_fp[1]->array(mfi);
+        amrex::Array4<amrex::Real> const& Ez_arr = E_fp[2]->array(mfi);
+        amrex::Array4<amrex::Real> const& Bt_arr = B_fp[1]->array(mfi);
+        amrex::Array4<amrex::Real> const& Bz_arr = B_fp[2]->array(mfi);
+
+        amrex::Array4<amrex::Real> const& pml_Et_arr = pml_E_fp[1]->array(mfi);
+        amrex::Array4<amrex::Real> const& pml_Bt_arr = pml_B_fp[1]->array(mfi);
+
+        // Get the tileboxes from Efield and Bfield so that they include the guard cells
+        // They are all the same, cell centered
+        amrex::Box tilebox = amrex::convert((*E_fp[1])[mfi].box(), E_fp[0]->ixType().toIntVect());
```

Comment:
If all the boxes are the same, are you choosing `*(E_fp[1])[mfi]` (so `1` instead of `0` or `2`) arbitrarily to extract the box? Just double checking to make sure I understand correctly.

Reply:
With RZ spectral, everything is cell centered so the box from any MultiFab can be used. The `1` is probably the result of a copy and paste. I'll change it to `0` to avoid confusion.